File: graph/scc.py
# SCC
def strongly_connected_component(G: list) -> list:
	n = len(G)
	G_rev = [[] for i in range(n)]
	for i in range(n):
		for v in G[i]:
			G_rev[v].append(i)
	vs = []
	visited = [False] * n
	def dfs(v):
		visited[v] = True
		for u in G[v]:
			if not visited[u]:
				dfs(u)
		vs.append(v)
	for i in range(n):
		if not visited[i]:
			dfs(i)
	rev_visited = [False] * n
	def rev_dfs(v):
		p.append(v)
		rev_visited[v] = True
		for u in G_rev[v]:
			if not rev_visited[u]:
				rev_dfs(u)
	res = []
	for v in vs[::-1]:
		if not rev_visited[v]:
			p = []
			rev_dfs(v)
			res.append(p)
	return res

File: Cpp/graph/scc.cpp
#include <algorithm>
#include <vector>

struct strongly_connected_components {
	public:
		strongly_connected_components() = default;
		explicit strongly_connected_components(int n) : G(n), G_rev(n), comp(n, -1), visited(n), rev_visited(n) {}

		void add_edge(int u, int v) {
			G[u].push_back(v);
			G_rev[v].push_back(u);
		}

		void build() {
			for (int i = 0; i < (int)G.size(); i++) {
				if (!visited[i]) dfs(i);
			}
			std::reverse(order.begin(), order.end());
			cnt = 0;
			for (int i : order) {
				if (!rev_visited[i]) rev_dfs(i, cnt++);
			}
		}

		int operator[](const int& i) {
			return comp[i];
		}

		int count() {
			return cnt;
		}

	private:
		std::vector<std::vector<int>> G, G_rev;
		std::vector<int> comp, order;
		std::vector<bool> visited, rev_visited;
		int cnt;

		void dfs(int u) {
			visited[u] = true;
			for (int v : G[u]) {
				if (!visited[v]) dfs(v);
			}
			order.push_back(u);
		}

		void rev_dfs(int u, int c) {
			rev_visited[u] = true;
			comp[u] = c;
			for (int v : G_rev[u]) {
			  if (!rev_visited[v]) rev_dfs(v, c);
			} 
		}
};